detection_rules/ml.py

```diff
@@ -0,0 +1,445 @@
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License
+# 2.0; you may not use this file except in compliance with the Elastic License
+# 2.0.
+
+"""Schemas and dataclasses for experimental ML features."""
+
+import io
+import zipfile
+from dataclasses import dataclass
+from functools import cached_property
+from pathlib import Path
+from typing import Dict, Literal, Optional
+
+import click
+import elasticsearch
+import json
+import requests
+from eql.table import Table
+from elasticsearch import Elasticsearch
+from elasticsearch.client import IngestClient, LicenseClient, MlClient
+
+from .eswrap import es_experimental
+from .ghwrap import Manifest, ReleaseManifest
+from .misc import client_error
+from .schemas import definitions
+from .utils import get_path
+
+
+ML_PATH = Path(get_path('machine-learning'))
+
+
+def info_from_tag(tag: str) -> (Literal['ml'], definitions.MachineLearningType, str, int):
+    try:
+        ml, release_type, release_date, release_number = tag.split('-')
+    except ValueError as exc:
+        raise ValueError(f'{tag} is not of valid release format: ml-type-date-number. {exc}')
+
+    return ml, release_type, release_date, int(release_number)
+
+
+class InvalidLicenseError(Exception):
+    """Invalid stack license for ML features requiring platinum or enterprise."""
+
+
+@dataclass
+class MachineLearningClient:
+    """Class for experimental machine learning release clients."""
+
+    es_client: Elasticsearch
+    bundle: dict
+
+    @cached_property
+    def model_id(self) -> str:
+        for name, data in self.bundle.items():
+            if Path(name).stem.lower().endswith('model'):
+                return data['model_id']
+
+    @cached_property
+    def bundle_type(self) -> str:
+        return self.model_id.split('_')[0].lower()
+
+    @cached_property
+    def ml_client(self) -> MlClient:
+        return MlClient(self.es_client)
+
+    @cached_property
+    def ingest_client(self) -> IngestClient:
+        return IngestClient(self.es_client)
+
+    @cached_property
+    def license(self) -> str:
+        license_client = LicenseClient(self.es_client)
+        return license_client.get()['license']['type'].lower()
+
+    def verify_license(self):
+        valid_license = self.license in ('platinum', 'enterprise')
+
+        if not valid_license:
+            err_msg = 'You must have a platinum or enterprise subscription in order to use these ML features'
```

Comment:
I'm not totally a fan of calling out subscriptions by name. My preference:
```suggestion
            err_msg = f'Your subscription level does not support Machine Learning. See ' \
              f'https://www.elastic.co/subscriptions for more information.'
```